src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java

```diff
@@ -289,6 +289,38 @@
  * }
  * </pre>
  * <p>
+ * To configure the check so that it limit IllegalType to validate field types only:
+ * </p>
+ * <pre>
+ * &lt;module name="IllegalType"&gt;
+ *   &lt;property name=&quot;illegalClassNames&quot; value=&quot;java.util.Optional&quot;/&gt;
+ *   &lt;property name=&quot;tokens&quot; value=&quot;VARIABLE_DEF&quot;/&gt;
+ * &lt;/module&gt;
+ * &lt;module name="SuppressionXpathSingleFilter"&gt;
+ *   &lt;property name=&quot;query&quot; value=&quot;//METHOD_DEF//*&quot;/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <pre>
+ * import java.util.Optional;
+ *
+ * public class Optional {
+ *
+ *   private Optional&lt;String&gt; field1; // violation, usage of type 'Optional' is not allowed
+ *   static int var = 4; // OK
+ *   protected notOptional&lt;String&gt; field2; // OK
+ *   public Optional&lt;String&gt; field3; // violation, usage of type 'Optional' is not allowed
+ *   Optional&lt;String&gt; field4; // violation, usage of type 'Optional' is not allowed
+ *
+ *   void foo() {
+ *     Optional&lt;String&gt; i; // OK
+ *   }
+ *
+ *   public &lt;T extends java.util.Optional&gt; void Method(T t){
+ *     Optional&lt;String&gt; i; // OK
+ *   }
+ * }
```

Comment:
@aryaniiit002 this code is still not compilable; why not use the code that is linked in the issue description as a starting point?

Reply:
code that is linked in the issue description

```
public class Test {
  public HashMap<String, String> function1() { // violation
    // code
  }

  private HashMap<String, String> function2() { // OK
    // code
  }

  protected HashMap<Integer, String> function3() { // violation
    // code
  }

  public static TreeMap<Integer, String> function4() { // violation
    // code
  }

}
```

```
Main.java:1: error: class Test is public, should be declared in a file named Test.java
public class Test {
       ^
Main.java:2: error: cannot find symbol
  public HashMap<String, String> function1() { // violation
         ^
  symbol:   class HashMap
  location: class Test
Main.java:6: error: cannot find symbol
  private HashMap<String, String> function2() { // OK
          ^
  symbol:   class HashMap
  location: class Test
Main.java:10: error: cannot find symbol
  protected HashMap<Integer, String> function3() { // violation
            ^
  symbol:   class HashMap
  location: class Test
Main.java:14: error: cannot find symbol
  public static TreeMap<Integer, String> function4() { // violation
                ^
  symbol:   class TreeMap
  location: class Test
5 errors
```
is this how code compiling looks like ? if yes see this please https://github.com/checkstyle/checkstyle/pull/9233#discussion_r569977282

Reply:
i think if https://github.com/checkstyle/checkstyle/pull/9233#discussion_r569977282 this is wrong then there are a lot of other code examples which are wrong or not compiling beside my code example

Reply:
i think if https://github.com/checkstyle/checkstyle/pull/9233#discussion_r569977282 this is wrong then there are a lot of other code examples which are wrong or not compiling beside my code example

this example is for 
`<property name="illegalClassNames" value="Boolean, Foo"/>
`just above my config
```
public class Main {

  public Set<Boolean> set; // violation
  public java.util.List<Map<Boolean, Foo>> list; // violation

  private void method(List<Foo> list, Boolean value) { // violation
    SomeType.<Boolean>foo(); // violation
    final Consumer<Foo> consumer = Foo<Boolean>::foo; // violation
  }

  public <T extends Boolean, U extends Serializable> void typeParam(T a) {} // violation

  public void fullName(java.util.ArrayList<? super Boolean> a) {} // violation

  public abstract Set<Boolean> shortName(Set<? super Boolean> a); // violation

  public Set<? extends Foo> typeArgument() { // violation
    return new TreeSet<Foo<Boolean>>();
  }

}
```

```
Main.java:3: error: cannot find symbol
  public Set<Boolean> set; // violation
         ^
  symbol:   class Set
  location: class Main
Main.java:4: error: cannot find symbol
  public java.util.List<Map<Boolean, Foo>> list; // violation
                        ^
  symbol:   class Map
  location: class Main
Main.java:4: error: cannot find symbol
  public java.util.List<Map<Boolean, Foo>> list; // violation
                                     ^
  symbol:   class Foo
  location: class Main
Main.java:6: error: cannot find symbol
  private void method(List<Foo> list, Boolean value) { // violation
                      ^
  symbol:   class List
  location: class Main
Main.java:6: error: cannot find symbol
  private void method(List<Foo> list, Boolean value) { // violation
                           ^
  symbol:   class Foo
  location: class Main
Main.java:11: error: cannot find symbol
  public <T extends Boolean, U extends Serializable> void typeParam(T a) {} // violation
                                       ^
  symbol:   class Serializable
  location: class Main
Main.java:15: error: cannot find symbol
  public abstract Set<Boolean> shortName(Set<? super Boolean> a); // violation
                                         ^
  symbol:   class Set
  location: class Main
Main.java:15: error: cannot find symbol
  public abstract Set<Boolean> shortName(Set<? super Boolean> a); // violation
                  ^
  symbol:   class Set
  location: class Main
Main.java:17: error: cannot find symbol
  public Set<? extends Foo> typeArgument() { // violation
         ^
  symbol:   class Set
  location: class Main
Main.java:17: error: cannot find symbol
  public Set<? extends Foo> typeArgument() { // violation
                       ^
  symbol:   class Foo
  location: class Main
Main.java:1: error: Main is not abstract and does not override abstract method shortName() in Main
public class Main {
       ^
Main.java:7: error: cannot find symbol
    SomeType.<Boolean>foo(); // violation
    ^
  symbol:   variable SomeType
  location: class Main
Main.java:8: error: cannot find symbol
    final Consumer<Foo> consumer = Foo<Boolean>::foo; // violation
          ^
  symbol:   class Consumer
  location: class Main
Main.java:8: error: cannot find symbol
    final Consumer<Foo> consumer = Foo<Boolean>::foo; // violation
                   ^
  symbol:   class Foo
  location: class Main
Main.java:8: error: method reference not expected here
    final Consumer<Foo> consumer = Foo<Boolean>::foo; // violation
                                   ^
Main.java:18: error: cannot find symbol
    return new TreeSet<Foo<Boolean>>();
               ^
  symbol:   class TreeSet
  location: class Main
Main.java:18: error: cannot find symbol
    return new TreeSet<Foo<Boolean>>();
                       ^
  symbol:   class Foo
  location: class Main
17 errors
```

Reply:
Please keep code compiling.
One day we will do validation of compilation automatically.

Reply:
> $ cat config.xml

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="IllegalType">
            <property name="illegalClassNames" value="java.util.Optional" />
            <property name="tokens" value="VARIABLE_DEF" />
        </module>
        <module name="SuppressionXpathSingleFilter">
          <property name="query" value="//METHOD_DEF//*"/>
        </module>
   </module>
</module>
```
> $ cat SomeClass.java

```
import java.util.Optional;

public class Main {

  private Optional<String> field1; // violation, usage of type 'Optional' is not allowed
  static int num = 4; // OK
  protected String field2; // OK
  public Optional<String> field3; // violation, usage of type 'Optional' is not allowed
  Optional<String> field4; // violation, usage of type 'Optional' is not allowed

  void foo() {
    Optional<String> i; // OK
  }
  public <T extends java.util.Optional> void Method(T t){
    Optional<String> i; // OK
  }
  public static void main(String args[]) {
    // ...implementation goes here...
  }
}
```
>  Starting audit...
```
[ERROR] /home/aryaniiit002/Desktop/checkstyle/.IgnoreFor_Checkstyle/SomeClass.java:5:11: Usage of type 'Optional' is not allowed. [IllegalType]
[ERROR] /home/aryaniiit002/Desktop/checkstyle/.IgnoreFor_Checkstyle/SomeClass.java:8:10: Usage of type 'Optional' is not allowed. [IllegalType]
[ERROR] /home/aryaniiit002/Desktop/checkstyle/.IgnoreFor_Checkstyle/SomeClass.java:9:3: Usage of type 'Optional' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 3 errors.
```
![Screenshot from 2021-02-08 10-29-01](https://user-images.githubusercontent.com/55039523/107177727-817a1680-69f8-11eb-98e5-cbdc4b2dbae0.png)
@nmancus1 @romani  pls check, this should work now.

Reply:
> One day we will do validation of compilation automatically.

please let me know if i can help anyway

> $ cat config.xml

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="IllegalType">
            <property name="illegalClassNames" value="java.util.Optional" />
            <property name="tokens" value="VARIABLE_DEF" />
        </module>
        <module name="SuppressionXpathSingleFilter">
          <property name="query" value="//METHOD_DEF//*"/>
        </module>
   </module>
</module>
```
> $ cat SomeClass.java

```
import java.util.Optional;

public class Main {

  private Optional<String> field1; // violation, usage of type 'Optional' is not allowed
  static int num = 4; // OK
  protected String field2; // OK
  public Optional<String> field3; // violation, usage of type 'Optional' is not allowed
  Optional<String> field4; // violation, usage of type 'Optional' is not allowed

  void foo() {
    Optional<String> i; // OK
  }
  public <T extends java.util.Optional> void Method(T t){
    Optional<String> i; // OK
  }
  public static void main(String args[]) {
    // ...implementation goes here...
  }
}
```
>  $ java -jar checkstyle-8.39-all.jar -c config.xml SomeClass.java

```
Starting audit...
[ERROR] /home/aryaniiit002/Desktop/checkstyle/.IgnoreFor_Checkstyle/SomeClass.java:5:11: Usage of type 'Optional' is not allowed. [IllegalType]
[ERROR] /home/aryaniiit002/Desktop/checkstyle/.IgnoreFor_Checkstyle/SomeClass.java:8:10: Usage of type 'Optional' is not allowed. [IllegalType]
[ERROR] /home/aryaniiit002/Desktop/checkstyle/.IgnoreFor_Checkstyle/SomeClass.java:9:3: Usage of type 'Optional' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 3 errors.
```
![Screenshot from 2021-02-08 10-29-01](https://user-images.githubusercontent.com/55039523/107177727-817a1680-69f8-11eb-98e5-cbdc4b2dbae0.png)
@nmancus1 @romani  pls check, this should work now.

Reply:
It is only necessary to share screenshots for generated site, code should always be CLI output. Please update this PR with what you have shown above.  You can remove `main()`, it doesn't really add anything to this example.

Reply:
done

Reply:
>please let me know if i can help anyway

Let's finish this PR to let you understand why are best practices, at least manually get to know them, after understanding comes we will see  how we can automatte it.

Reply:
Here is project for this https://github.com/checkstyle/checkstyle/wiki/Checkstyle-GSoC-2021-Project-Ideas#project-name-automate-verification-of-documentation-for-all-modules-and-generation-of-web-site-content-based-on-javadoc-of-modules
Such validation will be part of project